lib/yarp/ffi.rb

```diff
@@ -9,135 +12,200 @@ module YARP
-            arg_types = arg_types.map { |type| resolve_type(type.sub(/\w+$/, '')) }
-            return_type = resolve_type return_type
-            attach_function name, arg_types, return_type
-          end
-        end
+    # This object represents a yp_string_t. We only use it as an opaque pointer,
+    # so it doesn't have to be an FFI::Struct.
+    class YPString
+      attr_reader :pointer
+
+      def initialize(pointer)
+        @pointer = pointer
       end
-    end
 
-    load_exported_functions_from("yarp.h",
-      %w[yp_version yp_parse_serialize yp_lex_serialize])
-    load_exported_functions_from("yarp/util/yp_buffer.h",
-      %w[yp_buffer_init yp_buffer_free])
-    load_exported_functions_from("yarp/util/yp_string.h",
-      %w[yp_string_mapped_init yp_string_free yp_string_source yp_string_length yp_string_sizeof])
+      def source
+        LibRubyParser.yp_string_source(pointer)
+      end
+
+      def length
+        LibRubyParser.yp_string_length(pointer)
+      end
+
+      def read
+        source.read_string(length)
+      end
+    end
 
+    # This is the size of a yp_string_t. It is returned by the yp_string_sizeof
+    # function which we call once to ensure we have sufficient space for the
+    # yp_string_t FFI pointer.
     SIZEOF_YP_STRING = yp_string_sizeof
 
-    def self.pointer(size)
-      pointer = FFI::MemoryPointer.new(size)
+    # Yields a yp_string_t pointer to the given block.
+    def self.with_string(filepath, &block)
+      string = FFI::MemoryPointer.new(SIZEOF_YP_STRING)
+
       begin
-        yield pointer
+        raise unless yp_string_mapped_init(string, filepath)
+        yield YPString.new(string)
       ensure
-        pointer.free
+        yp_string_free(string)
+        string.free
       end
     end
-
-    def self.yp_string(&block)
-      pointer(LibRubyParser::SIZEOF_YP_STRING, &block)
-    end
   end
+
+  # Mark the LibRubyParser module as private as it should only be called through
+  # the YARP module.
   private_constant :LibRubyParser
 
-  VERSION = LibRubyParser.yp_version.to_s
+  # The version constant is set by reading the result of calling yp_version.
+  VERSION = LibRubyParser.yp_version.read_string
 
   def self.dump_internal(source, source_size, filepath)
-    buffer = LibRubyParser::Buffer.new
-    begin
-      raise unless LibRubyParser.yp_buffer_init(buffer)
-      metadata = nil
-      if filepath
-        metadata = [filepath.bytesize].pack('L') + filepath + [0].pack('L')
-      end
+    LibRubyParser.with_buffer do |buffer|
+      metadata = [filepath.bytesize, filepath.b, 0].pack("LA*L") if filepath
       LibRubyParser.yp_parse_serialize(source, source_size, buffer, metadata)
       buffer.to_ruby_string
-    ensure
-      LibRubyParser.yp_buffer_free(buffer)
-      buffer.pointer.free
     end
   end
   private_class_method :dump_internal
 
+  # Mirror the YARP.dump API by using the serialization API.
   def self.dump(code, filepath = nil)
```

Comment:
Maybe we should properly document these methods here, and then they would show up nicely in https://ruby.github.io/yarp/YARP.html like other methods defined in Ruby.
It certainly seems nicer than having to document in the middle of C code.
We might need to tell rdoc to ignore the C definitions somehow.